core/shared/src/main/scala/fs2/concurrent/Signal.scala

```diff
@@ -449,19 +488,23 @@ object SignallingMapRef {
               }
 
               def cleanup(id: Long): F[Unit] =
-                state.update { s =>
-                  val newListeners = s.listeners
-                    .get(k)
-                    .map(_ - id)
-                    .filterNot(_.isEmpty)
-                    .fold(s.listeners - k)(s.listeners.updated(k, _))
-                  s.copy(listeners = newListeners)
+                state.update { state =>
+                  state.keys.get(k).fold(state) { case KeyState(value, lastUpdate, listeners) =>
+                    val newListeners = listeners - id
+                    val newKeys =
+                      if (value.isEmpty && newListeners.isEmpty)
+                        state.keys - k // prevent memory leak
+                      else
+                        state.keys.updated(k, KeyState(value, lastUpdate, newListeners))
+                    state.copy(keys = newKeys)
+                  }
                 }
```

Comment:
Could we extract this lambda-block into a method of the `State` class? It only consists of non-effectful F code.